File: C_Tools/PcieEcamTool/PcieEcamTool.c
#include "stdio.h"
#include "sys/mman.h"
#include "string.h"
#include "stdbool.h"
#include "fcntl.h"
#include "unistd.h"

#define MaxBusNum                            256
#define MaxDevNum                            32
#define MaxFunNum                            8

#define PCI_MMIO_ADDRESS(Bus, Dev, Fun, Reg) (Bus<<20)+(Dev<<15)+(Fun<<12)+Reg
#define INTEL_PCIE_MMIO_BASE_ADDR            0xE0000000
#define AMD_PCIE_MMIO_BASE_ADDR              0xF0000000

#define BOUNDRY_46BIT                        0x3FFFFFFFFFFF // 64TB
#define MAX_MEM_ADDR                         ((unsigned long long)BOUNDRY_46BIT)

/**
  Print usage.
**/
void
PrintUsage (
  void
  )
{
  printf("PcieEcamTool:  usage\n");
  printf("  PcieEcamTool -h\n");
  printf("  PcieEcamTool -x\n");
  printf("  PcieEcamTool -s Bus:Dev.Fun\n");
  printf("  PcieEcamTool\n");
  printf("Parameter:\n");
  printf("  -h -help: show instructions\n");
  printf("  -s: specify a pcie device with the format of Bus:Dev.Fun.\n");
  printf("  -x: show pcie configuration space\n");
}

bool
IsDecimalDigitCharacter (
  char                    Char
) {
  return (bool) (Char >= '0' && Char <= '9');
}

char
Char16ToUpper (
  char                    Char
  )
{
  if (Char >= 'a' && Char <= 'z') {
    return (char) (Char - ('a' - 'A'));
  }

  return Char;
}

int
HexStrToUint16 (
  char              FirstChara,
  char              SecondChara
  )
{
  unsigned int      Data;

  if(FirstChara == SecondChara) {
    if (IsDecimalDigitCharacter(FirstChara)) {
      Data = FirstChara - '0';
    } else {
      if ((Char16ToUpper(SecondChara) - 'A') > 5) return -1;
      Data = Char16ToUpper(FirstChara) - 'A' + 10;
    }
  } else {
    if (IsDecimalDigitCharacter(FirstChara)) {
      Data = FirstChara - L'0';
    } else {
      if ((Char16ToUpper(FirstChara) - 'A') > 5) return -1;
      Data = Char16ToUpper(FirstChara) - 'A' + 10;
    }

    if (IsDecimalDigitCharacter(SecondChara)) {
      Data = (Data << 4) + (SecondChara - '0');
    } else {
      if ((Char16ToUpper(SecondChara) - 'A') > 5) return -1;
      Data = (Data << 4) + (Char16ToUpper(SecondChara) - 'A' + 10);
    }
  }

  return Data;
}

/**
  Print input parameters.
**/
int
ParseParameters (
  char                *InputStr,
  int                 *Bus,
  int                 *Dev,
  int                 *Fun 
  )
{
  unsigned short      Index;
  unsigned short      SemicolonPosition;
  unsigned short      DotPisition;
  unsigned short      InputStrLen;

  if (InputStr == NULL) {
    return -1;
  }

  for (Index = 0; InputStr[Index] != '\0'; Index ++) {
    if ((InputStr[Index] == ' ') || (InputStr[Index] == '\t')) return -1;
    if (InputStr[Index] == ':') SemicolonPosition = Index;
    if (InputStr[Index] == '.') DotPisition       = Index;
  }

  InputStrLen = Index;

  if ((SemicolonPosition > 2) || (DotPisition - SemicolonPosition > 3) || (InputStrLen - DotPisition > 2)) {
    return -1; 
  }

  *Bus = HexStrToUint16(InputStr[0], InputStr[SemicolonPosition - 1]);
  *Dev = HexStrToUint16(InputStr[SemicolonPosition + 1], InputStr[DotPisition - 1]);
  *Fun = HexStrToUint16(InputStr[DotPisition + 1], InputStr[InputStrLen - 1]);

  if (*Bus == -1 || *Dev == -1 || *Fun == -1) return -1;

  printf ("InputStrLen = %d, Bus = %x, Dev = %x, Fun = %x\n", InputStrLen, *Bus, *Dev, *Fun);

  return 0;
}

// This functino calculate a page aligned base and offset
int
AddrToMemPageBaseOffset (
  unsigned long          Addr,
  unsigned long          Length,
  size_t                 *Map_Size,
  unsigned long          *Base,
  unsigned long          *Offset
) {
  size_t                 SystemPageSize = sysconf(_SC_PAGE_SIZE);

  *Base   = Addr & (~(SystemPageSize -1));
  *Offset = Addr & (SystemPageSize -1);

  // *Map_Size is supposed to be at least the size of *Offset + Length, and should be the integer multiples of SystemPageSize;
  if ((*Offset + Length) > SystemPageSize) {
    *Map_Size = SystemPageSize * 2;
  } else {
    *Map_Size = SystemPageSize;
  }

  return 0;
}

int
MmioRead (
  unsigned long           PcieMmioAddr,
  unsigned int            Length,
  unsigned long           *Data
) {
  int                     Fd;
  volatile unsigned char  *Ptr_Mem;
  size_t                  Map_Size;
  unsigned long           Base;
  unsigned long           Offset;
  int                     ReturnValue;

  Fd                      = -1;
  Ptr_Mem                 = NULL;

  if (Data == NULL || (Length > sizeof(unsigned long long))) {
    return -1;
  }

  if (PcieMmioAddr > MAX_MEM_ADDR) {
    printf("ERROR: Memory address is greater than MAX_MEM_ADDR(0x%llx) supported.\n", MAX_MEM_ADDR);
    return -1;
  }

  // transform mapped physical address on mem page.
  ReturnValue = AddrToMemPageBaseOffset (PcieMmioAddr, Length, &Map_Size, &Base, &Offset);
  if (ReturnValue != 0) {
    return -1;
  }

  // Open mem device
  Fd = open("/dev/mem", O_RDWR);
  if (Fd < 0) {
    printf("ERROR: Failed to open %s.\n", "/dev/mem");
    ReturnValue = -1;

    goto Error;
  }

  Ptr_Mem = (volatile unsigned char*)mmap(NULL, Map_Size, PROT_READ, MAP_SHARED, Fd, Base);
  if (Ptr_Mem == (void *)-1) {
    printf("ERROR: Memory mapped failed.\n");
    ReturnValue = -1;
    goto Error;
  }

  if (Length == sizeof(unsigned char)) {
    *Data = *(volatile unsigned char*)(Ptr_Mem + Offset);
  } else if (Length == sizeof(unsigned short)) {
    *Data = *(volatile unsigned short*)(Ptr_Mem + Offset);
  } else if (Length == sizeof(unsigned int)) {
    *Data = *(volatile unsigned int*)(Ptr_Mem + Offset);
  } else if (Length == sizeof(unsigned long long)) {
    *Data = *(volatile unsigned long long*)(Ptr_Mem + Offset);
  }

Error:
  if (Fd >= 0) {
    close (Fd);
  }

  if (Ptr_Mem != NULL) {
    munmap ((void *)Ptr_Mem, Map_Size);
  }

  return ReturnValue;
}

unsigned char
MmioRead8 (
  unsigned long    Addr
) {
  unsigned long   Data;

  Data = 0;
  MmioRead (Addr, sizeof(unsigned char), &Data);

  return (unsigned char)Data;
}

unsigned short
MmioRead16 (
  unsigned long    Addr
) {
  unsigned long   Data;

  Data = 0;
  MmioRead (Addr, sizeof(unsigned short), &Data);

  return (unsigned short)Data;
}

unsigned int
MmioRead32 (
  unsigned long    Addr
) {
  unsigned long   Data;

  Data = 0;
  MmioRead (Addr, sizeof(unsigned int), &Data);

  return (unsigned int)Data;
}

void
ShowPciConfigSpace (
  unsigned long    PcieMmioAddr
) {
  unsigned short   Index;
  unsigned char    Data;

  for (Index = 0; Index <= 0xFFF; Index ++) {
    Data = MmioRead8 (PcieMmioAddr + Index);
    printf ("%02x ", Data);
    if ((Index + 1) % 16 == 0 ) printf ("\n");
    if (Index == 0x100 && MmioRead32 (PcieMmioAddr + Index) == 0) break;
  }
}

int main (
  int           Argc,
  char          *Argv[]
) {
  unsigned int  BusIndex;
  unsigned int  DevIndex;
  unsigned int  FunIndex;
  int           Bus;
  int           Dev;
  int           Fun;
  unsigned long PcieMmioAddr;
  unsigned int  VId;
  unsigned int  DId;
  int           ReturnValue;
  unsigned int  Index;
  bool          ShowRegs;

  PcieMmioAddr  = 0;
  VId           = 0;
  DId           = 0;
  Index         = 0;
  ShowRegs      = false;

  printf("Scanning Pcie devices by ECAM(MMIO) method:\n\n");

  // Parse input parameters.
  for (Index = 0; Index < Argc; Index ++) {
    if (strcmp(Argv[Index], "-h") == 0 || strcmp(Argv[Index], "-help") == 0) {
      PrintUsage();
      return 0;
    }
  }

  for (Index = 0; Index < Argc; Index ++) {
    if (strcmp(Argv[Index], "-x") == 0 || strcmp(Argv[Index], "-X") == 0) {
      ShowRegs = true;
    }
  }

  for (Index = 0; Index < Argc - 1; Index ++) {
    if (strcmp(Argv[Index], "-s") == 0 || strcmp(Argv[Index], "-S") == 0) {
      ReturnValue = ParseParameters(Argv[Index + 1], &Bus, &Dev, &Fun);
      if (ReturnValue == -1) {
        printf("Invalid Pci address.\n");
        return -1;
      }

      PcieMmioAddr = AMD_PCIE_MMIO_BASE_ADDR + PCI_MMIO_ADDRESS(Bus, Dev, Fun, 0);
      VId = MmioRead16 (PcieMmioAddr);
      if (VId == 0xffff) {
        printf("Invalid Pci address.\n");
        return -1;
      }

      DId = MmioRead16 (PcieMmioAddr + 2);
      printf("%02x:%02x.%x  VId = 0x%04x, DId = 0x%04x\n", Bus, Dev, Fun, VId, DId);
      if (ShowRegs) {
        ShowPciConfigSpace(PcieMmioAddr);
        printf("\n");
      }
    return 0;
    }
  }

  for (BusIndex = 0; BusIndex < MaxBusNum; BusIndex ++) {
    for (DevIndex = 0; DevIndex < MaxDevNum; DevIndex ++) {
      for (FunIndex = 0; FunIndex < MaxFunNum; FunIndex ++) {

        PcieMmioAddr = AMD_PCIE_MMIO_BASE_ADDR + PCI_MMIO_ADDRESS(BusIndex, DevIndex, FunIndex, 0);
        VId = MmioRead16 (PcieMmioAddr);
        if ((VId == 0xffff) && (FunIndex == 0)) {
          break;
        } else if (VId == 0xffff) {
          continue;
        }

        DId = MmioRead16 (PcieMmioAddr + 2);
        printf ("%02x:%02x.%x  VId = 0x%04x, DId = 0x%04x\n", BusIndex, DevIndex, FunIndex, VId, DId);
        if (ShowRegs) {
          ShowPciConfigSpace(PcieMmioAddr);
          printf ("\n");
        }
      }
    }
  }

  return 0;
}

File: Python_Tools/PcieEcamTool/PcieEcamTool.py
import os
import mmap
import sys
import getopt

MaxBusNum                         =   256
MaxDevNum                         =   32
MaxFuncNum                        =   8

AMD_PCIE_MMIO_BASE_ADDR           =  0xF0000000
INTEL_PCIE_MMIO_BASE_ADDR         =  0xE0000000
MAX_MEM_ADDR                      =  0x3FFFFFFFFFFF  # 64TB

def PCI_MMIO_ADDRESS (Bus, Dev, Func, Reg):
    return (Bus<<20)+(Dev<<15)+(Func<<12)+Reg

def MmioRead (Addr, Len):
    MAP_MASK = mmap.PAGESIZE - 1
    Fd = os.open("/dev/mem", os.O_RDWR | os.O_SYNC)
    Ptr_Mem = mmap.mmap(Fd, mmap.PAGESIZE, mmap.MAP_SHARED, mmap.PROT_WRITE | mmap.PROT_READ, offset = (Addr & ~MAP_MASK))
    Ptr_Mem.seek(Addr & MAP_MASK)
    Data_Arr = []
    for i in range(0, Len):
        Val = Ptr_Mem.read_byte()
        Data_Arr.append(Val)
        #print(hex(Val))
    Ptr_Mem.close()
    os.close(Fd)

    return Data_Arr

def ShowPciConfigSpace (Addr):
    for Index in range(0, 0x1000):
        Data = MmioRead(Addr + Index, 1)
        print ("%02x " %(Data[0]), end="")
        if ((Index + 1) % 16 == 0):
            print("")
        if (Index == 0x100 and MmioRead(Addr + Index, 1) == 0):
            break
    
def HelpMessage ():
    print("PcieEcamTool:  usage")
    print("  PcieEcamTool -h")
    print("  PcieEcamTool -x")
    print("  PcieEcamTool -s Bus:Dev.Func")
    print("  PcieEcamTool")
    print("Parameter:")
    print("  -h -help: show instructions")
    print("  -s: specify a pcie device with the format of Bus:Dev.Func")
    print("  -x: show pcie configuration space")

def HexStringToInt (str):
    Val = str.lower()
    return int(Val, 16)

if __name__ == "__main__":

    InputPcieDev = ""
    ShowRegs = False
    print("\nPython tool for doing Pcie devices by ECAM(MMIO):\n")

    # Parse input parameters.
    opts,args = getopt.getopt(sys.argv[1:], "hHxXs:S:", )
    for opt, arg in opts:
        if opt in ("-h", "-H"):
            HelpMessage()
            sys.exit()
        elif opt in ("-x", "-X"):
            ShowRegs = True
        elif opt in ("-s", "-S"):
            InputPcieDev = arg

    if InputPcieDev != "":
        parts = InputPcieDev.split(":")
        Bus = HexStringToInt(parts[0])
        subparts = parts[1].split(".")
        Dev = HexStringToInt(subparts[0])
        Func = int(subparts[1])
        # print (Bus, Dev, Func)
        if Bus < 0 or Bus >= MaxBusNum or Dev < 0 or Dev >= MaxDevNum or Func < 0 or Func >= MaxFuncNum:
            print("Input parameter for Bus:Dev.Func is wrong.\n")
            HelpMessage()
            sys.exit()
        else:
            PcieMmioAddr = AMD_PCIE_MMIO_BASE_ADDR + PCI_MMIO_ADDRESS(Bus, Dev, Func, 0)
            VId = MmioRead (PcieMmioAddr, 2)
            if VId[0] == 0xff and VId[1] == 0xff:
                print("Invalid Pci address")
                sys.exit()
            DId = MmioRead (PcieMmioAddr + 2, 2)
            print("%02x:%02x.%x  VId = 0x%02x%02x, DId = 0x%02x%02x" %(Bus, Dev, Func, VId[1], VId[0], DId[1], DId[0]))
            if ShowRegs == True:
                ShowPciConfigSpace(PcieMmioAddr)
                print("\n")
    else:
        for Bus in range(0, MaxBusNum):
            for Dev in range(0, MaxDevNum):
                for Func in range(0, MaxFuncNum):
                    PcieMmioAddr = AMD_PCIE_MMIO_BASE_ADDR + PCI_MMIO_ADDRESS(Bus, Dev, Func, 0)
                    VId = MmioRead (PcieMmioAddr, 2)
                    if VId[0] == 0xff and VId[1] == 0xff:
                        continue
                    DId = MmioRead (PcieMmioAddr + 2, 2)
                    print("%02x:%02x.%x  VId = 0x%02x%02x, DId = 0x%02x%02x" %(Bus, Dev, Func, VId[1], VId[0], DId[1], DId[0]))
                    if ShowRegs == True:
                        ShowPciConfigSpace(PcieMmioAddr)
                        print("\n")
